airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -335,11 +348,8 @@ def job_track_running(self) -> Union[AirbyteTracedException, requests.Response]:
             else:
                 # execute ERRORS scenario
                 self.on_job_with_errors(errors)
-        except (
-            ShopifyBulkExceptions.BulkJobBadResponse,
-            ShopifyBulkExceptions.BulkJobUnknownError,
-        ) as error:
-            return self.job_one_time_retry_error(response, error)
+        except self._retryable_errors as error:
+            return self.on_retryable_error(response, error)
```

Comment:
This feels odd to me. Given that `on_retryable_error` only perform the HTTP request again, we lose all the rest of the processing that `job_track_running` does so:
```
            errors = self.job_check_for_errors(response)
            if not errors:
                self.job_update_state(response)
                self.job_state_to_fn_map.get(self.job_state)(response=response)
                return response
            else:
                # execute ERRORS scenario
                self.on_job_with_errors(errors)
```

... won't be executed. Are these not important on retries? 

The same comment applies to `job_healthcheck`.

Reply:
It's expected! We want to be able to retry the failed request (latest one) because we track the job state for the current cycle iteration. Then, if everything is `healthy` - we reset the state and grab the new state from the incoming response for `RUNNING` job.

If the response for the RUNNING job could be validated or has some issues getting out the JSON () from the response - we should not go with the logic you mention but retry. Then, if the retried request is OK, we go with the rest of the logic.

I'm starting to think about how we can simplify the whole BULK flow, because I find it confusing for everyone around, including me, sometimes)

Reply:
Yeah, it is pretty complex...

To go back to my question above, let's say we have:

* `response = track_running_job.request(**status_args)` fails with an error within `_retryable_errors`, we will therefore call `self.on_retryable_error(response, error)`
* We call `response = track_running_job.request(**status_args)` again and it returns `{"errors": ["an error"]}`
    * Now, this seems very dangerous because `self.job_manager.job_check(response)` will return `None` and `process_bulk_results` will silently fail by not returning records
    
Hence, I think this way of retrying does not work. We should retry the whole method and not just the HTTP request

Reply:
> We call response = track_running_job.request(**status_args) again and it returns {"errors": ["an error"]}
Now, this seems very dangerous because self.job_manager.job_check(response) will return None and process_bulk_results will silently fail by not returning records


Based on the logic, we will `raise ShopifyBulkExceptions.BulkJobUnknownError()` with the details of what are these errors, like this:
```python
raise ShopifyBulkExceptions.BulkJobUnknownError(f"Could not validate the status of the BULK Job `123`. Errors: an error.")
```
In this scenario, the source will be stopped, we don't case about the response here, because we cannot move forward with the Errors, The platform will trigger the new attempt. When such things happen for Shopify - we should not just swallow that error but raise it. Didn't find anything dangerous retrying the latest request. 

It's clear if the Server returns 200 responses with the details about the actual Errors for the Bulk Job not being able to succeed. This is not user-actionable. Thus the error should be raised. We should check that and find the approach to fix it.

Reply:
I still think it does not work as expected. To leave the while loop in `job_check_state`,  we need to `self.job_completed() -> True`. In order to do that, `self.job_state_to_fn_map.get(self.job_state)(response=response)` needs to be called. It will not be called in the case of a retry as it is only called [here](https://github.com/airbytehq/airbyte/pull/37589/files#diff-69c119a9dc0a2f8a93bc200b41dfe1343d3ac94f117698a9442f30c5e2cadd60L330-L333). Can you add a test to validate this understanding?

Reply:
Changed the logic of checking the Bulk Job State, please check the code once again. Thank you.